### ISSUE TITLE
How to use extras.execute_values with a many-to-many data modifying CTE?

### ISSUE THREAD

**Issue by GollyJer** [psycopg/psycopg2#817]
My CTE based query takes some data, updates a couple tables, then updates a cross-reference table.
The problem is with `created_at` field.  It's of type `timestamp without time zone` in the database and when passed in from python it is `datetime`.  

When run as a single query it works. 👍

The problems show up when running in batches via the `execute many` function I get this error.
**column "created_at" is of type timestamp without time zone but expression is of type record**. 😕

I tried changing the line   
`SELECT twitter_id, created_at, content` 
to  
`SELECT twitter_id, created_at::timestamp, content`  
but then get the error  
**cannot cast type record to timestamp without time zone**. 🤔

Is this approach even possible via `execute_many()`?

Here's the full query. Line 9 is where the issue shows up.
```python
execute_values(
    self._cursor,
    """
    WITH input_data(twitter_id, created_at, content, search_hits) AS (
        VALUES (%s)
        )
    , twt AS (
        INSERT INTO tweet (twitter_id, created_at, content)
        SELECT twitter_id, created_at, content
        FROM input_data
            ON CONFLICT (twitter_id) DO UPDATE
            SET	twitter_id = excluded.twitter_id
        RETURNING id
        )
    , srch AS (
        INSERT INTO search (phrase)
        SELECT UNNEST(search_hits) FROM input_data
            ON CONFLICT (phrase) DO UPDATE
            SET phrase = excluded.phrase
        RETURNING id
        )
    INSERT INTO tweet_x_search (id_tweet, id_search)
    SELECT twt.id, id
        FROM srch
    ON CONFLICT DO NOTHING;
    """,
    tweets,
    page_size=1000,
)
```

Thanks for any help!
Jeremy

**Comment by dvarrazzo** [comment]
I'm sorry, this is a bug tracker, not a free on-line support. Write the mailing list for that.

**Comment by GollyJer** [comment]
Hi @dvarrazzo. Thanks for the response.
I'm trying to figure out if this **_is_** a bug or something I should be able to solve.

It seems type conversions aren't happening properly with CTEs when using the psycopg2 `extras.execute_values`.

If you're saying it should work and it's on my end that's great! I'll keep trying on my own.

Is that what you're saying?
Thanks.

**Comment by dvarrazzo** [comment]
It's your query that is emitting the wrong types. You can use the `pg_typeof()` database function to verify that.

**Comment by GollyJer** [comment]
Cool. I'll go down that path. Thanks!

**Comment by GollyJer** [comment]
@dvarrazzo I finally figured this out.  I'm putting the answer here posterity. 

To summarize the problem... CTEs don't know anything about types so when they pass values to the underlying queries you get the type errors.

The answer is to attach the types to the first row of the CTE values. At that point all values will be cast correctly when sent to the other queries.

I figured out how to game the system by creating an initial row of NULL values with the types defined. This tells the CTE the types.  Then simply ignore this "useless" first row when sending the data into the rest of the queries by using `OFFSET 1`.

Like this...
```python
execute_values(
    self._cursor,
    """
    WITH args (twitter_id, created_at, content, search_hits) AS (
        VALUES
        (NULL::text, NULL::timestamp, NULL::text, NULL::text[], NULL::text[]),
        %s
        OFFSET 1
    )
   <below here nothing changes>
    """,
    tweets,
    page_size=1000,
)
```

Best,
Jeremy
